photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,256 @@
+    each element in the input :math:`x` and target :math:`y`.
+
+
+    Args:
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. Default: ``'mean'``
+        data_range: The difference between the maximum and minimum of the pixel value,
+            i.e., if for image x it holds min(x) = 0 and max(x) = 1, then data_range = 1.
+            The pixel value interval of both input and output should remain the same.
+        scale_weights: Weights for different scales. Must contain 4 floating point values.
+        chromatic: Flag to use MS-GMSDc algorithm from paper.
+            It also evaluates chromatic components of the image. Default: True
+        beta1: Algorithm parameter. Weight of chromatic component in the loss.
+        beta2: Algorithm parameter. Small constant, see [1].
+        beta3: Algorithm parameter. Small constant, see [1].
+        
+    Reference:
+        [1] GRADIENT MAGNITUDE SIMILARITY DEVIATION ON MULTIPLE SCALES (2017)
+            http://www.cse.ust.hk/~psander/docs/gradsim.pdf
+    """
+
+    def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.,
+                 scale_weights: Optional[Union[Tuple[float], List[float]]] = None,
+                 chromatic: bool = False, beta1: float = 0.01, beta2: float = 0.32,
+                 beta3: float = 15.) -> None:
+        super().__init__()
+
+        # Generic loss parameters.
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.data_range = data_range
+        
+        # Values from the paper
+        if scale_weights is None:
+            self.scale_weights = torch.tensor([0.096, 0.596, 0.289, 0.019])
+        else:
+            self.scale_weights = torch.tensor(scale_weights)
+
+        self.chromatic = chromatic
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+            
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=self.scale_weights)
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+            
```

Comment:
Following #35, there should be an Assertion on the image sizes.

Reply:
Difference here is that for MS-SSIM minimal size is 32x32 and lot's of small datasets do not satisfy this condition. While for MS-GMSD minimal image size is 16x16. I doubt anyone will ever try to compute this metric between such small patches.

Reply:
The size depends on the number of scales used to calculate the measure. If user would like to use [.1]*10 weights, the size should be at least 2**len(weights) + 1.